File: customer-service/src/main/java/com/wks/servicemarketplace/customerservice/config/ApplicationParameters.java
package com.wks.servicemarketplace.customerservice.config;

import org.aeonbits.owner.Config;
import org.aeonbits.owner.ConfigFactory;

import java.net.URI;

@Config.LoadPolicy(Config.LoadType.MERGE)
@Config.Sources({"file:/etc/servicesmarketplace/application.properties",
        "classpath:application.properties",
        "system:env",
        "system:properties",
})
public interface ApplicationParameters extends Config {

    String DEFAULT_SERVER_HOST = "0.0.0.0";
    String DEFAULT_SERVER_PORT = "8080";
    String DEFAULT_OUTBOX_INTERVAL_MILLIS = "300";

    @Key("SERVER_HOST")
    @DefaultValue(DEFAULT_SERVER_HOST)
    String serverHost();

    @Key("SERVER_PORT")
    @DefaultValue(DEFAULT_SERVER_PORT)
    int serverPort();

    @Key("JDBC_URL")
    String jdbcUrl();

    @Key("JDBC_PASSWORD")
    String jdbcPassword();

    @Key("JDBC_USERNAME")
    String jdbcUsername();

    @Key("AMQP_URI")
    String amqpUri();

    @Key("AUTHSERVICE_HOST_URL")
    String authServiceBaseUrl();

    @Key("AUTHSERVICE_CLIENT_ID")
    String clientId();

    @Key("AUTHSERVICE_CLIENT_SECRET")
    String clientSecret();

    @Key("OUTBOX_INTERVAL_MILLIS")
    @DefaultValue(DEFAULT_OUTBOX_INTERVAL_MILLIS)
    Long outboxIntervalMillis();

    static ApplicationParameters load() {
        return ConfigFactory.create(ApplicationParameters.class);
    }
}

File: auth-service/src/main/kotlin/com/wks/servicemarketplace/authservice/config/ApplicationParameters.kt
package com.wks.servicemarketplace.authservice.config

import com.ufoscout.properlty.Properlty
import com.ufoscout.properlty.reader.EnvironmentVariablesReader
import com.ufoscout.properlty.reader.SystemPropertiesReader

interface ApplicationParameters {
    val serverHost: String
    val serverPort: Int
    val fusionAuthConfiguration: FusionAuthConfiguration
    val authServiceBaseUrl: String
    val jdbcUrl: String
    val jdbcUsername: String
    val jdbcPassword: String
    val amqpUri: String
    val amqpPrefetchCount: Int
    val clientId: String
    val clientSecret: String
    val retryAssignGroupIntervalMinutes: Long
    val outboxIntervalMillis: Long

    companion object {
        fun load() = DefaultApplicationParameters.load()
    }
}

data class FusionAuthConfiguration(private val provider: Properlty) {
    val serverUrl: String
        get() = provider["fusionauth.server.url"]!!
    val applicationId: String
        get() = provider["fusionauth.application.id"]!!
    val tenantId: String
        get() = provider["fusionauth.tenant.id"]!!
    val apiKey: String
        get() = provider["fusionauth.api.key"]!!
}

private class DefaultApplicationParameters(private val provider: Properlty) : ApplicationParameters {
    companion object {
        fun load(): ApplicationParameters {
            val provider = Properlty.builder()
                .caseSensitive(false)
                .add("file:/etc/servicesmarketplace/application.properties", ignoreNotFound = true)
                .add("classpath:application.properties", ignoreNotFound = true)
                .add(EnvironmentVariablesReader().replace("_", "."))
                .add(SystemPropertiesReader())
                .build()
            return DefaultApplicationParameters(provider)
        }
    }

    override val serverHost: String
        get() = provider["server.host", "0.0.0.0"]

    override val serverPort: Int
        get() = provider.getInt("server.port", 8082)

    override val fusionAuthConfiguration: FusionAuthConfiguration
        get() = FusionAuthConfiguration(provider)

    override val authServiceBaseUrl: String
        get() = provider["authservice.host.url"]!!

    override val clientId: String
        get() = provider["authservice.client.id"]!!

    override val clientSecret: String
        get() = provider["authservice.client.secret"]!!

    override val jdbcUrl: String
        get() = provider["jdbc.url"]!!

    override val jdbcUsername: String
        get() = provider["jdbc.username"]!!

    override val jdbcPassword: String
        get() = provider["jdbc.password"]!!

    override val amqpUri: String
        get() = provider["amqp.uri"]!!.let {
            it.replace("\u0020","")
        }

    override val amqpPrefetchCount: Int
        get() = provider.getInt("amqp.prefetchCount", 10)

    override val outboxIntervalMillis: Long
        get() = provider.getLong("app.outbox.intervalMillis", 500)

    override val retryAssignGroupIntervalMinutes: Long
        get() = provider.getLong("app.retryAssignGroup.intervalMinutes", 5)
}